src/box/alter.cc

```diff
@@ -4395,6 +4404,182 @@ on_replace_dd_schema(struct trigger * /* trigger */, void *event)
+	memset(def, 0, sizeof(*def));
+	if (tuple_field_uuid(tuple, BOX_GC_CONSUMERS_FIELD_UUID, &def->uuid) != 0)
+		return NULL;
+	if (tt_uuid_is_nil(&def->uuid)) {
+		diag_set(ClientError, ER_INVALID_UUID, tt_uuid_str(&def->uuid));
+		return NULL;
+	}
+	const char *mp_vclock =
+		tuple_field_with_type(tuple, BOX_GC_CONSUMERS_FIELD_VCLOCK,
+				      MP_MAP);
+	if (mp_vclock == NULL)
+		return NULL;
+	if (mp_decode_vclock_ignore0(&mp_vclock, &def->vclock) != 0) {
+		diag_set(ClientError, ER_INVALID_VCLOCK);
+		return NULL;
+	}
+	const char *opts = NULL;
+	if (tuple_field(tuple, BOX_GC_CONSUMERS_FIELD_OPTS) != NULL) {
+		opts = tuple_field_with_type(tuple, BOX_GC_CONSUMERS_FIELD_OPTS,
+					     MP_MAP);
+	}
+	if (gc_consumer_def_opts_decode(def, opts, region) != 0)
+		return NULL;
+	return def;
+}
+
+/**
+ * Data passed to transactional triggers of replace in _gc_consumers.
+ */
+struct gc_consumers_txn_trigger_data {
+	/*
+	 * Replica UUID. Is used instead of replica object because it can be
+	 * unregistered before the trigger is fired.
+	 */
+	struct tt_uuid uuid;
+	/*
+	 * New GC consumer, will be unregistered on rollback.
+	 */
+	struct gc_consumer *new_gc;
+};
+
+static int
+on_insert_update_dd_gc_consumers_rollback(struct trigger *trigger, void *)
+{
+	struct gc_consumers_txn_trigger_data *data =
+		(struct gc_consumers_txn_trigger_data *)trigger->data;
+	assert(data->new_gc != NULL);
+	gc_consumer_unregister(data->new_gc);
+	return 0;
+}
+
+static int
+on_replace_dd_gc_consumers_commit(struct trigger *trigger, void *)
+{
+	struct gc_consumers_txn_trigger_data *data =
+		(struct gc_consumers_txn_trigger_data *)trigger->data;
+	struct replica *replica = replica_by_uuid(&data->uuid);
+	/* Set new GC only if the replica is still registered. */
+	if (replica != NULL) {
+		/*
+		 * Drop the old gc consumer if any. If it's the first one,
+		 * unref gc_delay since this replica is already pinning xlogs.
+		 */
+		if (replica->gc != NULL)
+			gc_consumer_unregister(replica->gc);
+		else
+			gc_delay_unref();
+		replica->gc = data->new_gc;
+	} else {
+		if (data->new_gc != NULL)
+			gc_consumer_unregister(data->new_gc);
+	}
+	return 0;
+}
+
+/**
+ * The trigger invoked on replace in space _gc_consumers.
+ * Note that due to concurrent nature of transactions it is unsafe
+ * to modify replica->gc right here and in on_rollback, so it is
+ * modified only in on_commit triggers - it is safe because they
+ * are called in order of transactions' serialization.
+ */
+static int
+on_replace_dd_gc_consumers(struct trigger * /* trigger */, void *event)
+{
+	struct txn *txn = (struct txn *)event;
+	RegionGuard region_guard(&fiber()->gc);
+	struct txn_stmt *stmt = txn_current_stmt(txn);
+	struct tuple *old_tuple = stmt->old_tuple;
+	struct tuple *new_tuple = stmt->new_tuple;
+	struct gc_consumer_def *old_def = NULL;
+	struct gc_consumer_def *new_def = NULL;
+	struct tt_uuid *replica_uuid = NULL;
+	if (old_tuple != NULL) {
```

Comment:
What if old_tuple.uuid != new_tuple.uuid? We either should prohibit such replace or handle it properly: gc_consumer for old_tuple.uuid must be dropped.

Reply:
UUID is primary key in `_gc_consumers`, so they must be equal (that's how replace works).
Anyway, dropped an assertion here, just to make sure.